src/main/java/org/spongepowered/api/util/blockray/BlockRayHit.java

```diff
@@ -0,0 +1,141 @@
+package org.spongepowered.api.util.blockray;
+
+import com.flowpowered.math.GenericMath;
+import com.flowpowered.math.vector.Vector3d;
+import com.flowpowered.math.vector.Vector3i;
+import org.spongepowered.api.util.Direction;
+import org.spongepowered.api.world.extent.Extent;
+
+/**
+ * Represents a block hit by a ray. Stores more information than a regular location.
+ */
+public class BlockRayHit {
+
+    private final Extent extent;
+    private final double x, y, z;
+    private Vector3d position = null;
+    private final int xBlock, yBlock, zBlock;
+    private Vector3i blockPosition = null;
+    private final Direction face;
+
+    /**
+     * Constructs a new block ray hit from the extent that contains it, the coordinates
+     * and the face that was entered.
+     *
+     * @param extent The extent of the block
+     * @param x The x coordinate of the block
+     * @param y The y coordinate of the block
+     * @param z The x coordinate of the block
+     * @param face The entered face
+     */
+    public BlockRayHit(Extent extent, double x, double y, double z, Direction face) {
```

Comment:
Ok, I can add that. I'm also changing face to Vector3d because Direction doesn't provide values for up and down diagonals, which occur when rays enter corners (example, v = (1, 1, 1) and p = (0, 0, 0)).

In general I'm a bit unsure of how to handle sides and corners, I was thinking of averaging and normalizing the normal vectors of the faces that form it.